### PR TITLE
add line and fill vector components

### PR DESCRIPTION
**Pull request by freeman-lab**
This PR adds two new components `<Line/>` and `<Fill/>` for rendering vector layers from vector tiles.

First, some background. In a recent discussion we agreed that simpler building blocks for common vector layers would be nice to have, but we didn't want to build in any specific data dependencies e.g. by adding a `<Coastlines/>` component that required a specific hosted vector tile layer. However, with that decision, it remains cumbersome to quickly add simple line and fill vector layers. This recently became noticeable while trying to create simple demos.

As an alternative, this PR adds new components that render these layers from specified sources, so the dataset itself is completely outside the library. How? Rather than specif sources up front via the `style` object, the `<Line/>` and `<Fill/>` components have a `source` property. On construction, these components add the source _and_ add the layer to the map. An optional `id` prop allows explicitly using the same source across components, but by default we generate a random id.

The following renders the same coastline currently in the demo

```js
import { Map, Line } from '@carbonplan/maps'

<Map zoom={2} center={[0, 0]}>
  <Line
    color={theme.rawColors.primary} 
    source={'https://storage.googleapis.com/carbonplan-research/articles/offset-project-fire/basemap'}
    variable={'ne_10m_land'}
  />
</Map>
```

CleanShot 2021-09-26 at 11 58 26@2x

We copy the convention of `source` and `variable` to match the `Raster` component, though under the hood `variable` translates to a `sourceLayer` in mapbox terminology. It's easy to compose multiple components. In this example, we combine  three, and by specifying an `id` we ensure the common underlying source only loads once. (Note: we happen to be referencing a source here that has several variables included in it.)

```js
<Map zoom={2} center={[0, 0]}>
  <Fill
    color={theme.rawColors.muted} 
    source={'https://storage.googleapis.com/carbonplan-research/articles/offset-project-fire/basemap'}
    variable={'ne_10m_land'}
    id={'basemap'}
  />
  <Fill
    color={theme.rawColors.blue} 
    source={'https://storage.googleapis.com/carbonplan-research/articles/offset-project-fire/basemap'}
    variable={'ne_10m_lakes'}
    id={'basemap'}
  />
  <Line
    color={theme.rawColors.primary} 
    source={'https://storage.googleapis.com/carbonplan-research/articles/offset-project-fire/basemap'}
    variable={'ne_10m_land'}
    id={'basemap'}
  />
</Map>
```

CleanShot 2021-09-26 at 11 58 59@2x

I confirmed in testing that without specifying an `id` in the case above, the same tiles are loaded three times, whereas as written it only loads once. This seems like the desired behavior because, without explicitly specifying that multiple components use the same source, we should assume they are distinct.

A couple other small things:
- Because we've dropped `style` as a top-level prop, I renamed `containerStyle` -> `style` for the `Map`, `Mapbox`, and `Regl` components. Seems cleaner because it now better matches the built-in prop, but happy to discuss.
- I spent some time dialing the `useEffect` clean up behavior, and learned a couple things in the process. The main new finding is the ability to explicitly listen for an event that triggers when the map is removed, which avoids `map is undefined` errors in our `useEffect `cleanup functions.
- I avoided introducing a `theme-ui` dependency. It's perfectly easy to pass our theme colors in the `color` prop and we get theme dependence via the `useEffect` dependency on `color`. This feels like the right approach for now.
- We're explicitly assuming here that we're working with vector tiles in `pbf` format and standard `{z}/{x}/{y}` layout. We could relax that later, but it feels consistent with our opinionated assumptions around the `zarr` format and layout for raster data. In the short term, instead of relaxing, I think we should just aggressively validate.

### CODE REVIEW

File: src/fill.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { useEffect, useRef } from 'react'
+import { useMapbox } from './mapbox'
+import { updatePaintProperty } from './utils'
+import { v4 as uuidv4 } from 'uuid'
+
+const Fill = ({ source, variable, color, id, maxZoom = 5, opacity = 1 }) => {
+  const { map } = useMapbox()
+  const removed = useRef(false)
+
+  map.on('remove', () => {
```
Comment (katamartin) [inline src/fill.js:10]: 🤯

Comment (katamartin) [reply src/fill.js:10]: Should we be registering this listener just once on mount via `useEffect`?

**Review by katamartin — COMMENTED** [review]
This interface is super clean, love it!

It seems `Fill` and `Line` are share a lot of logic that could be consolidated, but we can leave for later if this is not straight-forward.

File: src/fill.js (hunk 1 above)
Comment (freeman-lab) [reply src/fill.js]: done!

**Pull request merged**